sdks/python/apache_beam/runners/worker/statesampler.pyx

```diff
@@ -71,19 +71,25 @@ cdef inline int64_t get_nsec_time() nogil:
       current_time.tv_nsec)
 
 
-class StateSamplerInfo(object):
-  """Info for current state and transition statistics of StateSampler."""
+class ExecutionStateSamplers(threading.local):
+  """ Per-thread state sampler. """
+  def __init__(self):
+    super(ExecutionStateSamplers, self).__init__()
+    self._current_sampler = None
 
-  def __init__(self, state_name, transition_count, time_since_transition):
-    self.state_name = state_name
-    self.transition_count = transition_count
-    self.time_since_transition = time_since_transition
+  def current_sampler(self):
+    return self._current_sampler
 
-  def __repr__(self):
-    return ('<StateSamplerInfo state: %s time: %dns transitions: %d>'
-            % (self.state_name,
-               self.time_since_transition,
-               self.transition_count))
+  def set_sampler(self, sampler):
+    self._current_sampler = sampler
+
+
+EXECUTION_STATE_SAMPLERS = ExecutionStateSamplers()
```

Comment:
<!--thread_id:cc_161077344_t; commit:bb69e90a90d3bae87cdb9f1620728b9735a803ac; resolved:0-->
<!--section:context-quote-->
> **robertwb** wrote:
> I'd really like to avoid more proliferation of global variables. This is somewhat needed for counters (as the user invokes counter operations without access to the underlying state) but we should be able to pass this state around explicitly.

<!--section:body-->
I'll document that we want to pass a tracker around as much as possible - but the aim of this global variable is to be able remove global state within MetricsEnvironment, and PerThreadLoggingContext, and have this be the canonical global state.

How does that sound?